groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -765,787 +1431,10708 @@ void Optional_DataImp<TYPE>::emplace(
+                BSLS_COMPILERFEATURES_FORWARD(ARGS_04, args_04),
+                BSLS_COMPILERFEATURES_FORWARD(ARGS_05, args_05));
+}
+#endif  // BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 5
+
+#if BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 6
+template <class TYPE, bool USES_BSLMA_ALLOC>
+template <class INIT_LIST_TYPE, class ARGS_01,
+                                class ARGS_02,
+                                class ARGS_03,
+                                class ARGS_04,
+                                class ARGS_05,
+                                class ARGS_06>
+inline
+optional<TYPE, USES_BSLMA_ALLOC>::optional(
+                               bsl::in_place_t,
+                               std::initializer_list<INIT_LIST_TYPE>      it,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_01) args_01,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_02) args_02,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_03) args_03,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_04) args_04,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_05) args_05,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_06) args_06)
+{
+    emplace(it, BSLS_COMPILERFEATURES_FORWARD(ARGS_01, args_01),
+                BSLS_COMPILERFEATURES_FORWARD(ARGS_02, args_02),
+                BSLS_COMPILERFEATURES_FORWARD(ARGS_03, args_03),
+                BSLS_COMPILERFEATURES_FORWARD(ARGS_04, args_04),
+                BSLS_COMPILERFEATURES_FORWARD(ARGS_05, args_05),
+                BSLS_COMPILERFEATURES_FORWARD(ARGS_06, args_06));
+}
+#endif  // BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 6
+
+#if BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 7
+template <class TYPE, bool USES_BSLMA_ALLOC>
+template <class INIT_LIST_TYPE, class ARGS_01,
+                                class ARGS_02,
+                                class ARGS_03,
+                                class ARGS_04,
+                                class ARGS_05,
+                                class ARGS_06,
+                                class ARGS_07>
+inline
+optional<TYPE, USES_BSLMA_ALLOC>::optional(
+                               bsl::in_place_t,
+                               std::initializer_list<INIT_LIST_TYPE>      it,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_01) args_01,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_02) args_02,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_03) args_03,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_04) args_04,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_05) args_05,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_06) args_06,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_07) args_07)
+{
+    emplace(it, BSLS_COMPILERFEATURES_FORWARD(ARGS_01, args_01),
+                BSLS_COMPILERFEATURES_FORWARD(ARGS_02, args_02),
+    
```

Comment:
Our member `swap` functions are different than free swap functions.  In particular, the member swap has a narrow contract - it is UB if allocators are different. The free swap function has a wide contract and will do more to avoid this UB. See `bdlb_nullablevalue` line 2280 for an example (quoted below)

```
template <class TYPE>
void bdlb::swap(NullableValue_WithAllocator<TYPE>& a,
                NullableValue_WithAllocator<TYPE>& b)
{
    if (a.d_allocator_p == b.d_allocator_p) {
        a.swap(b);

        return;                                                       // RETURN
    }

    NullableValue_WithAllocator<TYPE> futureA(b, a.d_allocator_p);
    NullableValue_WithAllocator<TYPE> futureB(a, b.d_allocator_p);

    futureA.swap(a);
    futureB.swap(b);
}
```

Reply:
this a curious one. I *can* do this for bsl::optional, but that means convincing LWG to do the same for pmr::optional. Let me poke around and see if that is achievable.

Reply:
Alisdair mentioned this http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0178r0.html
Perhaps it's worth involving him in this conversation? Maybe bunch this in the same email with `allocate_optional`?

Reply:
looking at the polls for that paper, it looks like LEWG was happy with that direction. That's enough to convince me LWG will be ok with that. I'll amend the free swap and the paper.